core/drivers/atmel_tcb.c

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Microchip
+ */
+
+#include <assert.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <io.h>
+#include <libfdt.h>
+#include <kernel/boot.h>
+#include <kernel/time_source.h>
+#include <matrix.h>
+#include <sama5d2.h>
+#include <tee_api_defines.h>
+
+#define TCB_CHAN(chan)		((chan) * 0x40)
+
+#define TCB_CCR(chan)		(0x0 + TCB_CHAN(chan))
+#define  TCB_CCR_SWTRG		0x4
+#define  TCB_CCR_CLKEN		0x1
+
+#define TCB_CMR(chan)		(0x4 + TCB_CHAN(chan))
+#define  TCB_CMR_WAVE		BIT32(15)
+#define  TCB_CMR_TIMER_CLOCK5	4
+#define  TCB_CMR_XC1		6
+#define  TCB_CMR_ACPA_SET	BIT32(16)
+#define  TCB_CMR_ACPC_CLEAR	SHIFT_U32(2, 18)
+
+#define TCB_CV(chan)		(0x10 + TCB_CHAN(chan))
+
+#define TCB_RA(chan)		(0x14 + TCB_CHAN(chan))
+#define TCB_RB(chan)		(0x18 + TCB_CHAN(chan))
+#define TCB_RC(chan)		(0x1c + TCB_CHAN(chan))
+
+#define TCB_IER(chan)		(0x24 + TCB_CHAN(chan))
+#define  TCB_IER_COVFS		0x1
+
+#define TCB_SR(chan)		(0x20 + TCB_CHAN(chan))
+#define  TCB_SR_COVFS		0x1
+
+#define TCB_IDR(chan)		(0x28 + TCB_CHAN(chan))
+
+#define TCB_BCR			0xc0
+#define  TCB_BCR_SYNC		0x1
+
+#define TCB_BMR			0xc4
+#define  TCB_BMR_TC1XC1S_TIOA0	SHIFT_U32(2, 2)
+
+#define TCB_WPMR		0xe4
+#define  TCB_WPMR_WAKEY		0x54494d
+
+static const char * const tcb_clocks[] = { "t0_clk", "gclk", "slow_clk" };
+static vaddr_t tcb_base;
+static uint32_t tcb_rate;
+
+static TEE_Result atmel_tcb_enable_clocks(const void *fdt, int node)
+{
+	unsigned int i = 0;
+	struct clk *clk = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	for (i = 0; i < ARRAY_SIZE(tcb_clocks); i++) {
+		res = clk_dt_get_by_name(fdt, node, tcb_clocks[i], &clk);
+		if (res)
+			return res;
+
+		clk_enable(clk);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result atmel_tcb_get_sys_time(TEE_Time *time)
+{
+	uint64_t cv0 = 0;
+	uint64_t cv1 = 0;
+
+	if (!tcb_base)
+		return TEE_ERROR_BAD_STATE;
```

Comment:
This means DT MUST define and enable the device, otherwise, OP-TEE time source if not available despite `REGISTER_TIME_SOURCE(atmel_tcb_time_source)`.

In that scope, I think `if (_fdt_get_status(fdt, node) != DT_STATUS_OK_SEC) return TEE_SUCCESS;` should rather panic at boot time or return some kind of error code.

Reply:
I actually want the clocks of all TCB blocks to be enabled whatever the status of the node. Maybe this could be check by some other means using a `late_initcall` for instance which would check the tcb_base variable.

Reply:
I added a boot_final call which check for tcb_base to be set.